File: src/eptools/matlab/test/binclass/eptest_binclass.m
% EPTOOLS Matlab Interface
% Example: Binary classification, coupled or factorized posterior
% mode.
% Can be configured with Gaussian or Laplace prior.
%
% Same data as glm-ie_v1.5/doc/classify_inference.m:
% UCI Adult (a9a): Dimension 123, first 30000 for testing, last 2561
% for training.

clear all, close all

%global debug_bmat; % DEBUG!
%global deb_ept_pi deb_ept_beta deb_ept_h deb_ept_rho; % DEBUG!
global debug_pydeb debug_pydeb_iter;
debug_pydeb = 0;
%debug_pydeb = 1;
debug_pydeb_iter = 0;

% Load and split into train and test set
load('/home/seeger/lhotse/src/eptools/matlab/test/binclass/classify.mat');
nte = 30000;      % say how many test examples we wich to keep
Bte = B(1:nte,:); B = B(nte+1:end,:); cte = c(1:nte);
c = c(nte+1:end); % split
% Remove variables not touched by any observations
ind = find(sum(B~=0,1));
B = B(:,ind); Bte = Bte(:,ind);
[q,n] = size(B);
% DEBUG:
%debug_bmat = [eye(n); B];

% Nearest neighbour classification
dp = sum(bsxfun(@minus, Bte, full(mean(B(c==+1,:)))).^2,2);  % dist. to pos. ctr
dm = sum(bsxfun(@minus, Bte, full(mean(B(c==-1,:)))).^2,2);  % dist. to neg. ctr
cc = 2*double(dp<dm)-1;
fprintf('nearest neighbor accuracy=%1.2f%%\n',100*sum(cte==cc)/ ...
	numel(cte))

% Setup
%imode = 'CoupParallel';   % Coupled, parallel updating
imode = 'CoupSequential'; % Coupled, sequential updating
%imode = 'Factorized';     % Factorized backbone
doLaplace = 1;            % Laplace prior
tau_lapl = 2/5;
%doLaplace = 0;            % Gaussian prior
%ssq_gauss = 25/4;
isFact = strcmp(imode,'Factorized');
if isFact
  seldampK = 5;           % Selective damping: K
end

% Coupling factors
if ~isFact
  model.matB = ept.MatContainer({ept.MatEye(n),ept.MatDef(B)});
  pmodel.matB = ept.MatDef(Bte);
else
  % Sparse matrices
  model.matB = [eye(n); B];
  pmodel.matB = Bte;
end
m = size(model.matB,1);
% Potential managers
if ~doLaplace
  model.potMan{1}.name = 'Gaussian';
  model.potMan{1}.size = n;
  model.potMan{1}.pars{1}=0;
  model.potMan{1}.pars{2}=ssq_gauss;
else
  model.potMan{1}.name = 'Laplace';
  model.potMan{1}.size = n;
  model.potMan{1}.pars{1}=0;
  model.potMan{1}.pars{2}=tau_lapl;
end
model.potMan{2}.name = 'Probit';
model.potMan{2}.size = q;
model.potMan{2}.pars{1}=c;
model.potMan{2}.pars{2}=0;
pmodel.potMan{1}.name = 'Probit';
pmodel.potMan{1}.size = nte;
pmodel.potMan{1}.pars{1}=cte;
pmodel.potMan{1}.pars{2}=0;
fprintf(1,'Prior:      %s\nLikelihood: %s\n', ...
	model.potMan{1}.name,model.potMan{2}.name);
% Initialization:
if ~doLaplace
  % NOTE: In factorized mode, OPTS.CAV_VAR is not needed, since all
  % Gaussian potentials have support 1
  [model,repr] = ept.init_ep(model,imode,'ADF');
else
  % Initialization as in glm-ie
  if ~isFact
    repr.epPi = ones(m,1);
    repr.epBeta = zeros(m,1);
    repr.epBeta(n+1:end) = 0.5*c;
  else
    % Factorized: Do not update on prior potentials in 1st sweep
    repr.epPi = zeros(nnz(model.matB),1);
    repr.epBeta = zeros(nnz(model.matB),1);
    repr.epPi(1:n) = 1;
  end
  repr = ept.refresh_repres(model,repr,imode);
end
% DEBUG:
%[repr.epPi(1:126) repr.epBeta(1:126)]
%pause;
if debug_pydeb
  if strcmp(imode,'CoupParallel')
    ep_pi = repr.epPi; ep_beta = repr.epBeta;
    lfact = repr.lFact; cvec = repr.cVec;
    mmeans = repr.mMeans; mvars = repr.mVars;
    save('eptb9_vars_it0.mat','ep_pi','ep_beta','lfact','cvec', ...
	 'mmeans','mvars','-v7');
  elseif strcmp(imode,'Factorized')
    ep_pi = repr.epPi; ep_beta = repr.epBeta;
    marg_pi = repr.mPi; marg_beta = repr.mBeta;
    save('eptb13_vars_it0.mat','ep_pi','ep_beta','marg_pi', ...
	 'marg_beta','-v7');
  end
end
% Predictions (before we start)
[h_q,rho_q,logz,h_p,rho_p] = ept.predict_ep(model,repr,imode,3, ...
					    pmodel);
acc = 100*sum(cte==sign(h_q))/numel(cte);
loglh = sum(logz)/numel(cte);
fprintf(1,['Initial predictions:\nAccuracy: %4.2f%%\n' ...
	   'Test set log likelihood: %.6f\n'],acc,loglh);

% EP inference
fprintf(1,'\nRunning EP inference (mode: %s)...\n',imode);
switch imode
 case 'CoupParallel'
  opts.maxit = 20;       % Max. number sweeps
  opts.deltaeps = 1e-4;  % Convergence threshold
  opts.damp = 0;         % No damping
  opts.caveps = 1e-5;
  opts.verbose = 1;
  %opts.test_model = pmodel;  % Test set predictions after each sweep
  t_start = tic;
  [model,repr,res,res_det] = ept.inf_ep(model,repr,imode,opts);
  t_inf = toc(t_start);
  fprintf(1,'\nTime for inference: %.6f\n',t_inf);
  if res.rstat==1
    fprintf(1,'\nDone MAXIT iterations.\n');
  else
    fprintf(1,'\nConverged to DELTAEPS accuracy.\n');
  end
  fprintf(1,['Number sweeps:   %d\n' ...
	     'Final delta:     %f\n' ...
	     'Number of skips: %d\n'],res.nit,res.delta,res.nskip);
 case 'CoupSequential'
  opts.maxit = 15;       % Max. number sweeps
  opts.deltaeps = 1e-4;  % Convergence threshold
  opts.damp = 0;         % No damping
  opts.caveps = 1e-5;
  opts.skipeps = 1e-7;
  opts.refresh = 1;      % Recompute repres. after each sweep
  opts.verbose = 1;
  %opts.test_model = pmodel;  % Test set predictions after each sweep
  t_start = tic;
  [model,repr,res,res_det] = ept.inf_ep(model,repr,imode,opts);
  t_inf = toc(t_start);
  fprintf(1,'\nTime for inference: %.6f\n',t_inf);
  if res.rstat==1
    fprintf(1,'\nDone MAXIT iterations.\n');
  else
    fprintf(1,'\nConverged to DELTAEPS accuracy.\n');
  end
  fprintf(1,['Number sweeps:   %d\n' ...
	     'Final delta:     %f\n' ...
	     'Skip histogram:  ['],res.nit,res.delta);
  fprintf(1,'%d ',res.nskip); fprintf(1,']\n');
 case 'Factorized'
  opts.maxit = 100;      % Max. number sweeps
  opts.deltaeps = 1e-4;  % Convergence threshold
  opts.damp = 0;         % No damping
  opts.piminthres = 1e-8;
  opts.refresh = 1;      % Recompute repres. after each sweep
  opts.verbose = 1;
  if ~doLaplace
    opts.skip_gauss = 1; % Skip Gaussian prior potentials
  else
    opts.upd_1stsweep = {'Probit'}; % Skip Laplace pots. 1st sweep
  end
  if exist('seldampK')
    % Selective damping mechanism
    repr.sd_numk = seldampK;
    model.matBInt = ept.bfact_intrepres(model.matB);
    irb = model.matBInt;
    if ~doLaplace
      % Exclude Gaussian prior potentials
      repr.sd_subind = ...
	  int32(ept.potman_filterpots(model.potMan, ...
				      {'Gaussian'})-1);
      repr.sd_subexcl = 1;
      [repr.sd_numvalid,repr.sd_topind,repr.sd_topval] = ...
	  eptools_fact_compmaxpi(n,m,irb.rowind,irb.colind, ...
				 irb.bvals,repr.epPi,repr.epBeta, ...
				 seldampK,repr.sd_subind, ...
				 repr.sd_subexcl);
    else
      [repr.sd_numvalid,repr.sd_topind,repr.sd_topval] = ...
	  eptools_fact_compmaxpi(n,m,irb.rowind,irb.colind, ...
				 irb.bvals,repr.epPi,repr.epBeta, ...
				 seldampK);
    end
    % DEBUG:
    %opts.debug_testmaxpi = 2; % Test after each sweep, full output
  end
  % DEBUG:
  %opts.debug_plotabsdiff = 1; % Plots after each sweep
  %opts.test_model = pmodel;  % Test set predictions after each sweep
  t_start = tic;
  [model,repr,res,res_det] = ept.inf_ep(model,repr,imode,opts);
  t_inf = toc(t_start);
  fprintf(1,'\nTime for inference: %.6f\n',t_inf);
  if res.rstat==1
    fprintf(1,'\nDone MAXIT iterations.\n');
  else
    fprintf(1,'\nConverged to DELTAEPS accuracy.\n');
  end
  fprintf(1,['Number sweeps:   %d\n' ...
	     'Final delta:     %f\n' ...
	     'Skip histogram:  ['],res.nit,res.delta);
  fprintf(1,'%d ',res.nskip); fprintf(1,']\n');
  if exist('seldampK')
    fprintf(1,'Select. damps:   %d\n',res.nsdamp);
  end
 otherwise
  error('Unknown IMODE');
end

% Predictions
[h_q,rho_q,logz,h_p,rho_p] = ept.predict_ep(model,repr,imode,3, ...
					    pmodel);
acc = 100*sum(cte==sign(h_q))/numel(cte);
loglh = sum(logz)/numel(cte);
fprintf(1,['\nFinal predictions:\nAccuracy: %4.2f%%\n' ...
	   'Test set log likelihood: %.6f\n'],acc,loglh);

File: python/test/binclass/eptest_binclass.py
#! /usr/bin/env python

# EPTOOLS Python Interface
# Example: Binary classification, coupled or factorized posterior
# mode.
# Can be configured with Gaussian or Laplace prior.
#
# Very similar to Matlab
#   testcode eptools/matlab/test/binclass/eptest_binclass.m
# Same data as glm-ie_v1.5/doc/classify_inference.m:
# Adult (a9a): Dimension 123, first 30000 for testing, last 2561
# for training.

import numpy as np
import scipy.sparse as ssp
import scipy.io  # DEBUG: Load Matlab files
import numbers
import time  # Profiling

import apbsint as abt

# Helper functions

# Main code

# Load dataset: CSV format. Inputs: Each row contains indices of nonzero
# features (value: 1).
# NOTE: Building a dense nested list is faster (by factor 10!) than
# building a ssp.lil_matrix row by row.
# NOTE: We remove 3 attributes (45, 116, 122), these are not touched by
# any patterns in the training set.
num_feat = n = 120  # After removing 3
tmat = []
fid = open('adult_a9a_inputs_comp.csv','r')
for line in fid:
    ind = [int(x) for x in line.split(',')]
    v = np.zeros(n+3,dtype=np.float64)
    v[ind] = 1.
    # Remove attributes 45, 116, 122
    tmat.append(list(np.hstack((v[:45], v[46:116], v[117:122]))))
fid.close()
num_cases = len(tmat)
print 'Dataset: Read %d cases.' % num_cases
inp_all = ssp.csr_matrix(tmat)
del tmat
num_test = 30000
num_train = num_cases-num_test
fid = open('adult_a9a_targets.csv','r')
targ_all = np.array([float(x) for x in fid.readline().split(',')],
                    dtype=np.float64)
fid.close()
if targ_all.size != num_cases:
    raise IndexError('Internal error: Wrong file size')

# Setup
#imode = 'CoupParallel'   # Coupled, parallel updating
#imode = 'CoupSequential' # Coupled, sequential updating
imode = 'Factorized'     # Factorized, sequential updating
is_fact = (imode == 'Factorized')
do_laplace = True        # Laplace prior
tau_lapl = 2./5.
#do_laplace = False       # Gaussian prior
#ssq_gauss = 25./4.
do_seldamp = True
#do_seldamp = False
seldamp_numk = 5
# DEBUG: Compare intermediate results of Matlab and Python code directly.
# We load Matlab variables from files written after init. (%d = 0) and after
# each sweep (%d = 1,2,...). Anything non-deterministic also loaded from there.
# NOTE: Implemented for 'Factorized' mode only.
#debug_matcomp_fname = '/home/seeger/exp/ept/eptb13_vars_it%d.mat'
try:
    debug_matcomp_fname
    do_debug_matcomp = True
except NameError:
    do_debug_matcomp = False

# Coupling factors
if not is_fact:
    bfct_test = abt.MatSparse(inp_all[:num_test,:].copy())
    bfct_train = abt.MatContainer([abt.MatEye(n),
                                   abt.MatSparse(inp_all[num_test:,:].copy())])
else:
    bfct_test = abt.MatFactorizedInf(inp_all[:num_test,:].copy())
    mx_tmp = ssp.vstack([ssp.eye(n,format='csr'), inp_all[num_test:,:]],
                        format='csr')
    bfct_train = abt.MatFactorizedInf(mx_tmp)
m = bfct_train.shape(0)
# Potential managers
if not do_laplace:
    pm_elem1 = abt.ElemPotManager('Gaussian',n,(0., ssq_gauss))
else:
    pm_elem1 = abt.ElemPotManager('Laplace',n,(0., tau_lapl))
pm_elem2 = abt.ElemPotManager('Probit',num_train,
                              (targ_all[num_test:].copy(), 0.))
pman_train = abt.PotManager((pm_elem1, pm_elem2))
pman_test = abt.PotManager(abt.ElemPotManager('Probit',num_test,
                                              (targ_all[:num_test].copy(), 0.)))
print 'Prior:      %s\nLikelihood: %s' % (pman_train.elem[0].name,
                                          pman_train.elem[1].name)
# Initialization of representation parameters
if not is_fact:
    model_train = abt.ModelCoupled(bfct_train,pman_train)
    model_test = abt.ModelCoupled(bfct_test,pman_test)
    repres = abt.RepresentationCoupled(bfct_train,
                                       keep_margs=(imode == 'CoupParallel'))
    if imode == 'CoupParallel':
        inf_driv = abt.EPCoupParallelInfDriver(model_train,repres)
    else:
        inf_driv = abt.EPCoupSequentialInfDriver(model_train,repres)
    if not do_laplace:
        inf_driv.init('ADF')
    else:
        # Same initialization as in glm-ie
        tvec = np.empty(m)
        tvec[:] = 1.
        repres.setpi(tvec)
        tvec[:n] = 0.
        tvec[n:] = 0.5*targ_all[num_test:]
        repres.setbeta(tvec)
        repres.refresh()
else:
    model_train = abt.ModelFactorized(bfct_train,pman_train)
    model_test = abt.ModelFactorized(bfct_test,pman_test)
    repres = abt.RepresentationFactorized(bfct_train)
    inf_driv = abt.EPFactorizedInfDriver(model_train,repres)
    if not do_laplace:
        # 'cav_var' argument does not matter, because coupling matrix for
        # Gaussian potentials is I
        inf_driv.init('ADF')
    else:
        # Below, we skip the prior potentials in the 1st sweep. We set their
        # pi's to 1, beta's to 0 (as in coupled mode and glm-ie)
        tvec = np.zeros(repres.size_pars())
        repres.setbeta(tvec)
        tvec[:n] = 1.
        repres.setpi(tvec)
        repres.refresh()
    if do_seldamp:
        if not do_laplace:
            # Exclude Gaussian potentials from selective damping mechanism
            # (they are not updated on anyway)
            sd_subind = pman_train.filterpots(set(['Gaussian']))
            repres.seldamp_reset(seldamp_numk,sd_subind,True)
        else:
            repres.seldamp_reset(seldamp_numk)
    # DEBUG
    if do_debug_matcomp:
        deb_mc = scipy.io.loadmat(debug_matcomp_fname % 0)
        deb_ep_pi = deb_mc['ep_pi'].ravel()
        deb_ep_beta = deb_mc['ep_beta'].ravel()
        deb_marg_pi = deb_mc['marg_pi'].ravel()
        deb_marg_beta = deb_mc['marg_beta'].ravel()
        print ('DEBUG[%d]: df(ep_pi)=%.4e, df(ep_beta)=%.4e, ' +
               'df(m_ep)=%.4e, df(m_beta)=%.4e') % \
               (0, abt.helpers.maxreldiff(repres.ep_pi,deb_ep_pi),
                abt.helpers.maxreldiff(repres.ep_beta,deb_ep_beta),
                abt.helpers.maxreldiff(repres.marg_pi,deb_marg_pi),
                abt.helpers.maxreldiff(repres.marg_beta,deb_marg_beta))
# Initial test set prediction
opts = abt.helpers.Struct()
opts.imode = imode
opts.ptype = 3
(h_q, rho_q, logz, h_p, rho_p) = inf_driv.predict(model_test,opts)
acc = 100.*float((np.sign(h_q)==targ_all[:num_test]).sum())/num_test
loglh = logz.sum()/num_test
print ('\nInitial predictions:\nAccuracy: %4.2f%%\n'
       'Test set log likelihood: %.6f') % (acc, loglh)

# EP inference
print '\nRunning EP inference (mode: %s)...' % imode
if imode == 'CoupParallel':
    opts = abt.helpers.Struct()
    opts.imode = imode
    opts.maxit = 20       # Max. number sweeps
    #opts.maxit = 40       # Max. number sweeps
    opts.deltaeps = 1e-4  # Convergence threshold
    opts.damp = 0.        # No damping
    opts.caveps = 1e-5
    opts.verbose = 1
    opts.res_det = True
    opts.bc_testmodel = model_test
    t_start = time.time()
    (res, res_det) = inf_driv.inference(opts)
    t_stop = time.time()
    print 'Time(inference): %.6fs' % (t_stop-t_start)
    if res.rstat==1:
        print '\nDone MAXIT iterations.'
    else:
        print '\nConverged to DELTAEPS accuracy.'
    print ('Number sweeps:   %d\n' +
           'Final delta:     %f\n' +
           'Number of skips: %d') % (res.nit,res.delta,res.nskip)
elif imode == 'CoupSequential':
    opts = abt.helpers.Struct()
    opts.imode = imode
    opts.maxit = 15       # Max. number sweeps
    opts.deltaeps = 1e-4  # Convergence threshold
    opts.damp = 0.        # No damping
    opts.caveps = 1e-5
    opts.skipeps = 1e-7
    opts.refresh = True   # Recompute repres. after each sweep
    opts.verbose = 1
    opts.res_det = True
    #opts.bc_testmodel = model_test
    t_start = time.time()
    (res, res_det) = inf_driv.inference(opts)
    t_stop = time.time()
    print 'Time(inference): %.6fs' % (t_stop-t_start)
    if res.rstat==1:
        print '\nDone MAXIT iterations.'
    else:
        print '\nConverged to DELTAEPS accuracy.'
    print ('Number sweeps:   %d\n' +
           'Final delta:     %f') % (res.nit,res.delta)
    print 'Skip histogram: ', res.nskip
elif imode == 'Factorized':
    opts = abt.helpers.Struct()
    opts.imode = imode
    opts.maxit = 100      # Max. number sweeps
    opts.deltaeps = 1e-4  # Convergence threshold
    opts.damp = 0.        # No damping
    opts.piminthres = 1e-7
    opts.refresh = True   # Recompute repres. after each sweep
    opts.verbose = 1
    opts.res_det = True
    if not do_laplace:
        # Gaussian potentials have coupling factor I, so EP updates do
        # not change anything. Skip them
        opts.skip_gauss = True
    else:
        # Skip updates for Laplace prior potentials in 1st sweep
        opts.upd_1stsweep = set(['Probit'])
    if do_debug_matcomp:
        opts.deb_matcomp_fname = debug_matcomp_fname
    opts.bc_testmodel = model_test
    t_start = time.time()
    (res, res_det) = inf_driv.inference(opts)
    t_stop = time.time()
    print 'Time(inference): %.6fs' % (t_stop-t_start)
    if res.rstat==1:
        print '\nDone MAXIT iterations.'
    else:
        print '\nConverged to DELTAEPS accuracy.'
    print ('Number sweeps:   %d\n' +
           'Final delta:     %f') % (res.nit,res.delta)
    print 'Skip histogram:  ', res.nskip
    if do_seldamp:
        print 'Select. damps:   %d' % res.nsdamp
else:
    raise ValueError("IMODE: Unknown value '" + imode + "'")

# Predictions
opts = abt.helpers.Struct()
opts.imode = imode
#opts.ptype = 0
opts.ptype = 3
#h_q = inf_driv.predict(model_test,opts)
(h_q, rho_q, logz, h_p, rho_p) = inf_driv.predict(model_test,opts)
acc = 100.*float((np.sign(h_q)==targ_all[:num_test]).sum())/num_test
loglh = logz.sum()/num_test
print ('\nFinal predictions:\nAccuracy: %4.2f%%\n'
       'Test set log likelihood: %.6f') % (acc, loglh)
